src/fsharp/FSharp.Core/array.fs

```diff
@@ -483,19 +483,45 @@ namespace Microsoft.FSharp.Collections
                             count <- count + 1
             Array.subUnchecked 0 count res
 
-
         [<CompiledName("Filter")>]
-        let filter f (array: _[]) = 
-            checkNonNull "array" array
-            let res = Array.zeroCreateUnchecked array.Length 
-            let mutable count = 0
-            for x in array do                 
-                if f x then 
-                    res.[count] <- x
-                    count <- count + 1
-            Array.subUnchecked 0 count res
-
+        let filter f (array: _[]) =         
+            checkNonNull "array" array                    
+            let mutable i = 0    
+            while i < array.Length && not (f array.[i]) do
+                i <- i + 1
+            
+            if i <> array.Length then                    
+                let mutable element = array.[i]
+                let chunk1 : 'T[] = Array.zeroCreateUnchecked (((array.Length-i) >>> 2) + 1)
```

Comment:
@jackmott 
I'm not sure about the chunking heuristic.  Just out of curiosity ... what would the results look like ... if you took the above optimisation and applied it to the existing algorithm?